### PR TITLE
Add support for Kotlin 1.4's functional interfaces

### PR DESCRIPTION
**Pull request by ZacSweers**
This adds a new `KModifier.FUN` and orders it alongside `enum` and `annotation` (figured this was the most conventional spot).

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 1]
```diff
@@ -743,6 +746,14 @@ class TypeSpec private constructor(
         }
       }
 
+      if (isFunInterface) {
+        // Note: Functional interfaces can contain any number of non-abstract functions.
+        val abstractFunSpecs = funSpecs.filter { ABSTRACT in it.modifiers }
+        check(abstractFunSpecs.size == 1) {
+          "Functional interfaces must have exactly one abstract function. Contained ${abstractFunSpecs.size}: ${abstractFunSpecs.map { it.name }}"
```
Comment (JakeWharton) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt:753]: Wrap

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 2]
```diff
@@ -846,6 +846,52 @@ class TypeSpecTest {
...
+        |""".trimMargin())
+  }
+
+  @Test fun funInterface_empty_shouldError() {
+    try {
+      TypeSpec.funInterfaceBuilder("Taco")
+          .build()
+      fail()
+    } catch (e: IllegalStateException) {
+      assertThat(e).hasMessageThat().contains("Functional interfaces must have exactly one abstract function. Contained 0")
```
Comment (JakeWharton) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt:875]: Wrap

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 3]
```diff
@@ -846,6 +846,52 @@ class TypeSpecTest {
...
+          .addFunction(FunSpec.builder("fun1")
+              .addModifiers(ABSTRACT)
+              .build())
+          .addFunction(FunSpec.builder("fun2")
+              .addModifiers(ABSTRACT)
+              .build())
+          .build()
+      fail()
+    } catch (e: IllegalStateException) {
+      assertThat(e).hasMessageThat().contains("Functional interfaces must have exactly one abstract function. Contained 2")
```
Comment (JakeWharton) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt:891]: Wrap

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt (hunk 1 above)
Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: 2df8283

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt (hunk 2 above)
Comment (ZacSweers) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: 2df8283

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt (hunk 3 above)
Comment (ZacSweers) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: 2df8283

**Comment by ZacSweers** [comment]
Shoot, I just thought to ask this - thoughts on `isFunInterface` vs `isFunctionalInterface`?

**Comment by JakeWharton** [comment]
I prefer the latter, given the choice

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 4]
```diff
@@ -846,6 +846,56 @@ class TypeSpecTest {
...
+        |  fun sam()
+        |
+        |  fun notSam() {
+        |  }
+        |}
+        |""".trimMargin())
+  }
+
+  @Test fun funInterface_empty_shouldError() {
+    try {
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt:870]: `assertThrows`

**Comment by Egorand** [comment]
@ZacSweers wanna update the README too?

**Comment by ZacSweers** [comment]
Yeah I'll do a follow-up PR for the above 3 bits

**Pull request merged**